pyked/converters.py

```diff
@@ -0,0 +1,741 @@
+from .chemked import ChemKED, DataPoint
+from .utils import units as unit_registry
+from ._version import __version__
+
+
+# Exceptions
+class ParseError(Exception):
+    """Base class for errors."""
+    pass
+
+class KeywordError(ParseError):
+    """Raised for errors in keyword parsing."""
+
+    def __init__(self, *keywords):
+        self.keywords = keywords
+
+    def __str__(self):
+        return repr('Error: {}.'.format(self.keywords[0]))
+
+class MissingElementError(KeywordError):
+    """Raised for missing required elements."""
+
+    def __str__(self):
+        return repr('Error: Required element {} is missing.'.format(
+            self.keywords[0]))
+
+class MissingAttributeError(KeywordError):
+    """Raised for missing required attribute."""
+
+    def __str__(self):
+        return repr('Error: Required attribute {} of {} is missing.'.format(
+            self.keywords))
+
+class UndefinedKeywordError(KeywordError):
+    """Raised for undefined keywords."""
+
+    def __str__(self):
+        return repr('Error: Keyword not defined: {}'.format(self.keywords[0]))
+
+
+def get_file_metadata(root):
+    """Read and parse ReSpecTh XML file metadata (file author, version, etc.)
+
+    Args:
+        root (`etree.Element`): Root of ReSpecTh XML file
+
+    Returns:
+        properties (`dict`): Dictionary with file metadata
+    """
+    properties = {}
+
+    properties['file-author'] = {'name': ''}
+    try:
+        properties['file-author']['name'] = root.find('fileAuthor').text
+    except AttributeError:
+        raise AttributeError('Error: no fileAuthor given')
+
+    if properties['file-author']['name'] == '':
+        raise AttributeError('Error: no fileAuthor given')
+
+    # Default version is 0
+    version = '0'
+    elem = root.find('fileVersion')
+    if elem is None:
+        print('Warning: no fileVersion given')
+    else:
+        try:
+            version = elem.find('major').text + '.' + elem.find('minor').text
+        except AttributeError:
+            print('Error: missing fileVersion major/minor')
+            raise
+    properties['file-version'] = int(float(version))
+
+    # Default ChemKED version
+    properties['chemked-version'] = __version__
+
+    return properties
+
+
+def get_reference(root):
+    """Read reference info from root of ReSpecTh XML file.
+
+    Args:
+        root (`etree.Element`): Root of ReSpecTh XML file
+
+    Returns:
+        properties (`dict`): Dictionary with reference information
+    """
+    reference = {}
+    elem = root.find('bibliographyLink')
+    if elem is None:
+        raise AttributeError('Error: missing bibliographyLink')
+
+    # Try to get reference info via DOI
+    try:
+        reference['doi'] = elem.attrib['doi']
+        ref = None
+        try:
+            ref = habanero.Crossref().works(ids=reference['doi'])['message']
+        except (HTTPError, habanero.RequestError):
+            print('DOI not found')
+            raise KeyError
```

Comment:
this `KeyError` is raised so that the outer `except` catches it, and switches to parsing the `preferredKey` attribute for reference information. `doi` isn't a required element of ReSpecTh files, so it shouldn't be a proper error—only if both `doi` *and* `preferredKey` are missing.